pkg/sync/rpc/server.go

```diff
@@ -445,16 +447,77 @@ func (s *SyncAgentServer) FilesSync(ctx context.Context, req *enginerpc.FilesSyn
 		}
 	}()
 
+	if req.LocalSync != nil {
+		err := s.fileSyncLocal(ctx, req)
+		if err == nil {
+			return &emptypb.Empty{}, nil
+		}
+
+		logrus.WithError(err).Warn("Falling back to remote sync")
+	}
+
+	return &emptypb.Empty{}, s.fileSyncRemote(ctx, req)
+}
+
+func (s *SyncAgentServer) fileSyncLocal(ctx context.Context, req *enginerpc.FilesSyncRequest) error {
+	var targetPaths []string
+	var err error
+
+	log := logrus.WithFields(logrus.Fields{
+		"sourcePath": req.LocalSync.SourcePath,
+		"targetPath": req.LocalSync.TargetPath,
+	})
+
+	log.Info("Syncing files locally")
+
+	// Defer function to handle cleanup of files if an error occurs
+	defer func() {
+		if err == nil {
+			log.Info("Done syncing files locally")
+		} else {
+			log.WithError(err).Warn("Failed to sync files locally, reverting changes")
+
+			for _, targetPath := range targetPaths {
+				if removeErr := os.Remove(targetPath); removeErr != nil && removeErr != os.ErrNotExist {
+					log.WithError(removeErr).Warnf("Failed to remove file %v", targetPath)
+				}
+			}
+		}
+	}()
+
+	for _, info := range req.SyncFileInfoList {
+		sourcePath := filepath.Join("/host", req.LocalSync.SourcePath, info.FromFileName)
+		targetPath := filepath.Join("/host", req.LocalSync.TargetPath, info.ToFileName)
+		targetPaths = append(targetPaths, targetPath)
+
+		log.Tracef("Copying file %v to %v", sourcePath, targetPath)
+
+		err = sparse.SyncLocalFile(sourcePath, targetPath)
+		if err != nil {
+			return err
+		}
+
+		err := lhio.CheckIsFileSizeSame(sourcePath, targetPath)
+		if err != nil {
+			log.WithError(err).Tracef("Failed to check file size for file %v", sourcePath)
```

Comment:
Why not use Errorf?
Should it error out if the sizes are mismatched?

Reply:
I've updated to use Warnf as it seems more appropriate since we are only logging the error without returning it. What do you think? I am also fine changing it to Errorf if you think that is more appropriate.